// @grammyjs/storage-file/dist/cjs/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileAdapter = void 0;
const deps_node_js_1 = require("./deps.node.js");
class FileAdapter {
    /**
     * @constructor
     * @param {opts} options options
     * @param {opts.dirName} options.dirName - name of directory where files should be stored. Defaults to `sessions`.
     * @param {opts.serializer} options.serializer
     * serializer of file. Default `JSON.stringify(input, null, '\t')`.
     *
     * @param {opts.deserializer} options.deserializer
     * deserializer of file. Default `JSON.parse(input)`.
     */
    constructor(opts = {}) {
        var _a, _b, _c;
        this.folderPath = deps_node_js_1.path.resolve((0, deps_node_js_1.cwd)(), (_a = opts === null || opts === void 0 ? void 0 : opts.dirName) !== null && _a !== void 0 ? _a : 'sessions');
        this.serializer = (_b = opts.serializer) !== null && _b !== void 0 ? _b : ((input) => JSON.stringify(input, null, '\t'));
        this.deserializer = (_c = opts.deserializer) !== null && _c !== void 0 ? _c : ((input) => JSON.parse(input));
        deps_node_js_1.fs.ensureDirSync(this.folderPath);
    }
    resolveSessionPath(key) {
        const subFolder = key.substr(-2);
        return deps_node_js_1.path.resolve(this.folderPath, subFolder, `${key}.json`);
    }
    async findSessionFile(key) {
        try {
            return await deps_node_js_1.fs.readFile(this.resolveSessionPath(key));
        }
        catch {
            return null;
        }
    }
    async read(key) {
        const file = await this.findSessionFile(key);
        if (!file) {
            return undefined;
        }
        return this.deserializer(file);
    }
    async write(key, value) {
        const fullPath = this.resolveSessionPath(key);
        const fileName = `${key}.json`;
        const folderPath = fullPath.substring(0, fullPath.length - fileName.length);
        await deps_node_js_1.fs.ensureDir(folderPath);
        await deps_node_js_1.fs.writeFile(fullPath, this.serializer(value));
    }
    async delete(key) {
        try {
            await deps_node_js_1.fs.remove(this.resolveSessionPath(key));
        }
        catch { }
    }
}
exports.FileAdapter = FileAdapter;
